From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Minecrell <minecrell@minecrell.net>
Date: Thu, 21 Sep 2017 16:33:12 +0200
Subject: [PATCH] Allow plugins to use SLF4J for logging

SLF4J is a commonly used abstraction for various logging frameworks
such as java.util.logging (JUL) or Log4j. Currently, plugins are
required to do all their logging using the provided JUL logger.
This is annoying for plugins that target multiple platforms or when
using libraries that log messages using SLF4J.

Expose SLF4J as optional logging API for plugins, so they can use
it without having to shade it in the plugin and going through
several layers of logging abstraction.

diff --git a/build.gradle.kts b/build.gradle.kts
index 48fff1c1dd7f77f8451b541d75fdb7adeda8312d..b82dc32ddd22195830badf674396dfebf4775942 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -40,6 +40,8 @@ dependencies {
     apiAndDocs(libs.adventure.text.serializer.legacy)
     apiAndDocs(libs.adventure.text.serializer.plain)
     apiAndDocs(libs.adventure.text.logger.slf4j)
+    api(libs.log4j.api)
+    api(libs.slf4j.api)
 
     implementation(libs.asm)
     implementation(libs.asm.commons)
@@ -108,7 +110,9 @@ tasks.withType<Javadoc> {
         "https://jd.advntr.dev/text-serializer-gson/${libs.versions.adventure.get()}/",
         "https://jd.advntr.dev/text-serializer-legacy/${libs.versions.adventure.get()}/",
         "https://jd.advntr.dev/text-serializer-plain/${libs.versions.adventure.get()}/",
-        "https://jd.advntr.dev/text-logger-slf4j/${libs.versions.adventure.get()}/"
+        "https://jd.advntr.dev/text-logger-slf4j/${libs.versions.adventure.get()}/",
+        "https://javadoc.io/doc/org.slf4j/slf4j-api/${libs.versions.slf4j.api.get()}/",
+        "https://javadoc.io/doc/org.apache.logging.log4j/log4j-api/${libs.versions.log4j.api.get()}/"
         // Paper end
     )
     options.tags("apiNote:a:API Note:")
diff --git a/src/main/java/org/bukkit/plugin/Plugin.java b/src/main/java/org/bukkit/plugin/Plugin.java
index 8c76716249e44ed8bf6be94c1f5c7b6d9bb35be2..4eb639fbb46a0848be207149ea433455550fae1c 100644
--- a/src/main/java/org/bukkit/plugin/Plugin.java
+++ b/src/main/java/org/bukkit/plugin/Plugin.java
@@ -198,6 +198,22 @@ public interface Plugin extends TabExecutor {
     }
     // Paper end
 
+    // Paper start - Add SLF4J/Log4J loggers
+    @NotNull
+    default org.slf4j.Logger getSLF4JLogger() {
+        return org.slf4j.LoggerFactory.getLogger(getLogger().getName());
+    }
+
+    /**
+     * @deprecated use {@link #getSLF4JLogger()}
+     */
+    @Deprecated
+    @NotNull
+    default org.apache.logging.log4j.Logger getLog4JLogger() {
+        return org.apache.logging.log4j.LogManager.getLogger(getLogger().getName());
+    }
+    // Paper end
+
     /**
      * Returns the name of the plugin.
      * <p>
